--- /home/haohuang/backup/aosp13-TP1A.221005.002/frameworks/opt/telephony/src/java/com/android/internal/telephony/GsmCdmaPhone.java	2023-03-07 20:02:54.824250827 +0000
+++ /home/haohuang/aosp13-TP1A.221005.002/frameworks/opt/telephony/src/java/com/android/internal/telephony/GsmCdmaPhone.java	2023-03-06 21:01:43.601671542 +0000
@@ -86,6 +86,7 @@
 import android.text.TextUtils;
 import android.util.Log;
 import android.util.Pair;
+import android.os.Binder;
 
 import com.android.ims.ImsManager;
 import com.android.internal.annotations.VisibleForTesting;
@@ -135,6 +136,9 @@
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
+import com.android.internal.telephony.util.ProcessUtil;
+import com.android.internal.telephony.RILDefender;
+
 /**
  * {@hide}
  */
@@ -1616,6 +1620,23 @@
             ResultReceiver wrappedCallback)
             throws CallStateException {
 
+        // RILDefender: Detect voice calls from untrusted apps
+        // obtain caller's PID and name
+        int callerPid = Binder.getCallingPid();
+        String callerName = ProcessUtil.getAppNameByPID(getContext(), callerPid).trim();
+        Rlog.d(LOG_TAG, "Dialing request received from pid = " + callerPid + " " + callerName);
+
+        // check source
+        if (callerName != null && RILDefender.trustedDialerAppName.contains(callerName)) {
+            Rlog.d(LOG_TAG, "Setting validCall = true");
+            RILDefender.validCall = true;
+        } else {
+            Rlog.d(LOG_TAG, "Voice call initated from an untrusted app " + callerName);
+            RILDefender.validCall = false;
+            // do not let suspecious app do dialing
+            //return null;
+        }
+
         // Need to make sure dialString gets parsed properly
         String newDialString = PhoneNumberUtils.stripSeparators(dialString);
 
