--- /home/haohuang/backup/aosp13-TP1A.221005.002/frameworks/base/telephony/java/com/android/internal/telephony/gsm/SmsMessage.java	2023-03-07 20:13:32.994785100 +0000
+++ /home/haohuang/aosp13-TP1A.221005.002/frameworks/base/telephony/java/com/android/internal/telephony/gsm/SmsMessage.java	2023-03-06 21:01:43.645671982 +0000
@@ -1104,7 +1104,7 @@
      * Returns the TP-Data-Coding-Scheme byte, for acknowledgement of SMS-PP download messages.
      * @return the TP-DCS field of the SMS header
      */
-    int getDataCodingScheme() {
+    public int getDataCodingScheme() {
         return mDataCodingScheme;
     }
 
@@ -1638,6 +1638,10 @@
                 (mProtocolIdentifier == 0x7f || mProtocolIdentifier == 0x7c);
     }
 
+    boolean isClassZero() {
+        return messageClass == MessageClass.CLASS_0;
+    }
+
     public int getNumOfVoicemails() {
         /*
          * Order of priority if multiple indications are present is 1.UDH,
