--- /home/haohuang/backup/aosp10-QP1A.190711.019/frameworks/base/telephony/java/android/telephony/SignalStrength.java	2023-02-17 21:57:24.193267782 +0000
+++ /home/haohuang/aosp10-QP1A.190711.019/frameworks/base/telephony/java/android/telephony/SignalStrength.java	2023-02-17 19:32:59.772714434 +0000
@@ -195,6 +195,18 @@
         return mLte;
     }
 
+    // print signals in dbm as strings
+    public String getSignalStrings() {
+        List<CellSignalStrength> css = getCellSignalStrengths();
+        StringBuilder str = new StringBuilder();
+        str.append(getPrimary().getClass().getSimpleName());
+        for (CellSignalStrength cs: css) {
+            str.append("\t");
+            str.append(cs.getDbm());
+        }
+        return str.toString();
+    }
+
     /**
      * Returns a List of CellSignalStrength Components of this SignalStrength Report.
      *
